arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/RaiseAccumulate.kt

```diff
@@ -502,6 +502,70 @@ public inline fun <Error, A, B, C, D, E, F, G, H, I, J> Raise<NonEmptyList<Error
   return block(unbox(a), unbox(b), unbox(c), unbox(d), unbox(e), unbox(f), unbox(g), unbox(h), unbox(i))
 }
 
+@RaiseDSL
+public inline fun <Error, A> Raise<Error>.forEachAccumulating(
+  iterable: Iterable<A>,
+  combine: (Error, Error) -> Error,
+  @BuilderInference block: RaiseAccumulate<Error>.(A) -> Unit
+): Unit = forEachAccumulating(iterable.iterator(), combine, block)
+
+@RaiseDSL
+public inline fun <Error, A> Raise<Error>.forEachAccumulating(
+  sequence: Sequence<A>,
+  combine: (Error, Error) -> Error,
+  @BuilderInference block: RaiseAccumulate<Error>.(A) -> Unit
+): Unit = forEachAccumulating(sequence.iterator(), combine, block)
+
+@RaiseDSL
+public inline fun <Error, A> Raise<Error>.forEachAccumulating(
+  iterator: Iterator<A>,
+  combine: (Error, Error) -> Error,
+  @BuilderInference block: RaiseAccumulate<Error>.(A) -> Unit
+): Unit = recover({
+  for (item in iterator) {
+    block(RaiseAccumulate(this), item)
+  }
+}) { firstErrors ->
+  var error = firstErrors.reduce(combine)
+  for (item in iterator) {
+    recover({
+      block(RaiseAccumulate(this), item)
+    }) {
+      error = combine(error, it.reduce(combine))
+    }
+  }
+  raise(error)
+}
```

Comment:
I find this code quite hard to understand. I would prefer to have some code in the style of the previous `mapOrAccumulate`

```suggestion
): Unit {
  var error: Any? = EmptyValue
  for (item in iterable) {
    fold<NonEmptyList<Error>, B, Unit>(
      { transform(RaiseAccumulate(this), item) },
      { errors -> error = combine(error, errors.reduce(combine), combine) },
      { }
    )
  }
  return if (error === EmptyValue) results else raise(unbox<Error>(error))
}
```

Reply:
I wanted to get rid of `EmptyValue` because it added too much complexity to the code. Also, I had hoped to provide some way to access whether or not an error has happened yet, so that `mapOrAccumulate` can stop storing successful values when it no longer needs to. Finally, I liked the fact that for the success case, which should probably be the most likely case, the code only does one allocation for the first `recover`, while the old code would allocate one everytime. It's a premature optimization yes, but I figured the code isn't too horrible. WDYT? I'll revert it back for now because this is orthogonal to the PR